core/src/main/java/com/linecorp/armeria/server/http/auth/Authorizer.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.http.auth;
+
+import java.util.concurrent.CompletableFuture;
+
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * Determines whether a given {@code data} is authorized for the service registered in.
+ * {@code ctx} can be used for storing authorization information about the request for use in
+ * business logic. {@code data} is usually a {@link com.linecorp.armeria.common.http.HttpRequest}
+ * or token extracted from it.
+ *
+ * <p>Returns a {@link CompletableFuture} that will resolve to {@code true} if the request is
+ * authorized of {@code false} otherwise. If the future resolves exceptionally, the request
+ * will not be authorized.
+ */
+@FunctionalInterface
+public interface Authorizer<T> {
+    CompletableFuture<Boolean> authorize(T data, ServiceRequestContext ctx);
```

Comment:
Moved some javadoc here. TBH, it feels better for FunctionalInterface to only require javadoc in one place since it's just a single function, but I guess that's not standardized anywhere.